### DISCUSSION TITLE
react-native subscription promo how they work in Android  ?

### DISCUSSION THREAD

**Discussion by muhammadsaadullah1** [hyochan/react-native-iap#2563] [Q&A]
SO I have react-native subscription code 
` const result = await IAP.requestSubscription({skus:['my sku'],subscriptionOffers: [{sku: 'my sku', offerToken}]});` 
it works fine  but i have generated  google play store promos  and   want the user to redeem them . 

How should I ? 
Thanks in advance I have been working on it for days  and I got no documentation that is helping with this.

**Comment by Esarhaddon** [comment]
Hey, @khan8901, I just finished refactoring my company's Play Store in app purchase logic to use `react-native-iap` and had I hard time understanding how to correctly implement promos/offers as well. Here's how to get it working (assuming you've configured everything in the Play Console, which it sounds like you have):

1) The first thing is to realize that an offer token tells the Android billing library which specific offer to subscribe the user to. Maybe that's already obvious to you, but at first I thought `offerToken` was just a generic token required to make a purchase on Android. However, if you want to a user to subscribe using a particular offer, you have to call `requestSubscription` with the correct `offerToken`.

1) To find the correct `offerToken` you'll want to know the product id, base plan id, and offer id associated with the offer, which you'll be able to find in the Play Console. If you're not already familiar with the "subscription => base plan => offer" hierarchy on Android, you can find an explanation of that [here](https://support.google.com/googleplay/android-developer/answer/12154973#).

1) Once you have that information then in your app you can do `results = await IAP.getSubscriptions({skus: [<your skus here>]})` and then iterate over `results`. Each item in `results` represents a subscription product and will have a `subscriptionOfferDetails` property, which will be an array of all the offers the user is eligible for that are associated with that particular subscription. Each offer will have `basePlanId`, `offerId`, and `offerToken` properties. Using the product id, base plan id, and offer id from step 2 you'll be able to find the offer token you need.
   > NOTE: By default a base plan is associated with an offer that will have an `offerId` of `null`. That will be the only offer in `subscriptionOfferDetails` associated with a given base plan if you haven't configured any offers for the base plan. See the [Google Play Billing Client Docs](https://developer.android.com/reference/com/android/billingclient/api/ProductDetails.SubscriptionOfferDetails#getOfferId()) for `SubscriptionOfferDetails`.

1) Once you have the offer token you need, then you should be able to call `IAP.requestSubscription({skus:['my sku'],subscriptionOffers: [{sku: 'my sku', offerToken}]})` to subscribe the user to that offer.
   > NOTE: Testing free trials can be tricky on Android as there's no way to reset trial eligibility for a test user like there is on iOS, but my coworker has had success using an offer with the "developer determined" eligibility criteria.

Hope that helps.

**Comment by MarwaKhaledMohamed** [comment]
Hi  @muhammadsaadullah1,

Were you able to get the promo code working on Android? I have the same issue — I created a promo, but in both the testing and production environments , the 'Redeem Code' option doesn't appear.
